pkg/whip/server.go

```diff
@@ -0,0 +1,225 @@
+package whip
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/livekit/ingress/pkg/config"
+	"github.com/livekit/ingress/pkg/errors"
+	"github.com/livekit/protocol/logger"
+	"github.com/livekit/protocol/rpc"
+	"github.com/livekit/protocol/utils"
+	"github.com/livekit/psrpc"
+)
+
+const (
+	sdpResponseTimeout = 5 * time.Second
+	rpcTimeout         = 5 * time.Second
+)
+
+type WHIPServer struct {
+	onPublish func(streamKey, resourceId, sdpOffer string) error
+	handlers  sync.Map
+	rpcClient rpc.IngressHandlerClient
+}
+
+type handler struct {
+	sdpChan chan<- sdpRes
+}
+
+type sdpRes struct {
+	sdp string
+	err error
+}
+
+func NewWHIPServer(rpcClient rpc.IngressHandlerClient) *WHIPServer {
+	return &WHIPServer{
+		rpcClient: rpcClient,
+	}
+}
+
+func (s *WHIPServer) handleError(err error, w http.ResponseWriter) {
+	var psrpcErr psrpc.Error
+	switch {
+	case errors.As(err, &psrpcErr):
+		w.WriteHeader(psrpcErr.ToHttp())
+		w.Write([]byte(psrpcErr.Error()))
+	case err == nil:
+		// Nothing, we already responded
+	default:
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+}
+
+func (s *WHIPServer) Start(conf *config.Config, onPublish func(streamKey, resourceId, sdpOffer string) error) error {
+	s.onPublish = onPublish
+
+	r := mux.NewRouter()
+
+	r.HandleFunc("/{app}", func(w http.ResponseWriter, r *http.Request) {
+		var err error
+		defer func() {
+			s.handleError(err, w)
+		}()
+
+		bearer := r.Header.Get("Authorization")
+		if !strings.HasPrefix(bearer, "Bearer ") {
+			err = psrpc.NewErrorf(psrpc.NotFound, "missing stream name in authorization header")
+			return
+		}
+
+		streamKey := strings.TrimPrefix(bearer, "Bearer ")
+
+		err = s.handleNewWhipClient(w, r, streamKey)
+	}).Methods("POST")
+
+	r.HandleFunc("/{app}/{stream_key}", func(w http.ResponseWriter, r *http.Request) {
+		var err error
+		defer func() {
+			s.handleError(err, w)
+		}()
+
+		streamKey := mux.Vars(r)["stream_key"]
+
+		err = s.handleNewWhipClient(w, r, streamKey)
+	}).Methods("POST")
+
+	r.HandleFunc("/{app}", func(w http.ResponseWriter, r *http.Request) {
+		setCORSHeaders(w, r, false)
+	}).Methods("OPTIONS")
+
+	r.HandleFunc("/{app}/{stream_key}", func(w http.ResponseWriter, r *http.Request) {
+		setCORSHeaders(w, r, false)
+	}).Methods("OPTIONS")
+
+	// End
+	r.HandleFunc("/{app}/{stream_key}/{resource_id}", func(w http.ResponseWriter, r *http.Request) {
+		var err error
+		defer func() {
+			s.handleError(err, w)
+		}()
+
+		vars := mux.Vars(r)
+		resourceID := vars["resource_id"]
+
+		logger.Infow("handling WHIP delete request", "resourceID", resourceID)
+
+		req := &rpc.DeleteWHIPResourceRequest{
+			ResourceId: resourceID,
+		}
+
+		_, err = s.rpcClient.DeleteWHIPResource(context.Background(), resourceID, req, psrpc.WithRequestTimeout(5*time.Second))
+	}).Methods("DELETE")
+
+	// Trickle, ICE Restart
```

Comment:
does WHIP support trickle ICE?

Reply:
It does in the client -> server path. See section 4.1 of the RPC (https://www.ietf.org/archive/id/draft-ietf-wish-whip-08.txt). I do not know if any client currently support it.